subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2106 @@
+	{"AT+CLIP", bt_hfp_ag_clip_handler},
+};
+
+NET_BUF_POOL_FIXED_DEFINE(ag_pool, CONFIG_BT_HFP_AG_TX_BUF_COUNT,
+			  BT_RFCOMM_BUF_SIZE(BT_HF_CLIENT_MAX_PDU),
+			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static struct bt_hfp_ag bt_hfp_ag_pool[CONFIG_BT_MAX_CONN];
+
+static struct bt_hfp_ag_cb *bt_ag;
+
+/* Sent but not acknowledged TX packets with a callback */
+static struct bt_ag_tx ag_tx[CONFIG_BT_HFP_AG_TX_BUF_COUNT * 2];
+static K_FIFO_DEFINE(ag_tx_free);
+static K_FIFO_DEFINE(ag_tx_notify);
+
+struct k_thread ag_thread;
+static K_KERNEL_STACK_MEMBER(ag_thread_stack, CONFIG_BT_HFP_AG_THREAD_STACK_SIZE);
+static k_tid_t ag_thread_id;
+
+static struct bt_conn *bt_hfp_ag_create_sco(struct bt_hfp_ag *ag);
+
+static void bt_hfp_ag_set_call_state(struct bt_hfp_ag *ag, bt_hfp_call_state_t state);
+
+static void bt_hfp_ag_set_state(struct bt_hfp_ag *ag, bt_hfp_state_t state)
+{
+	ag->state = state;
+
+	switch (state) {
+	case BT_HFP_DISCONNECTED:
+		if (bt_ag && bt_ag->disconnected) {
+			bt_ag->disconnected(ag);
+		}
+		break;
+	case BT_HFP_CONNECTING:
+		break;
+	case BT_HFP_CONFIG:
+		break;
+	case BT_HFP_CONNECTED:
+		if (bt_ag && bt_ag->connected) {
+			bt_ag->connected(ag);
+		}
+		break;
+	case BT_HFP_DISCONNECTING:
+		break;
+	default:
+		LOG_WRN("no valid (%u) state was set", state);
+		break;
+	}
+}
+
+static void bt_hfp_ag_set_call_state(struct bt_hfp_ag *ag, bt_hfp_call_state_t state)
+{
+	ag->call_state = state;
+
+	switch (state) {
+	case BT_HFP_CALL_TERMINATE:
+		atomic_clear(ag->flags);
+		k_work_cancel_delayable(&ag->deferred_work);
+		break;
+	case BT_HFP_CALL_OUTGOING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_OUTGOING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_INCOMING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_INCOMING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ALERTING:
+		k_work_reschedule(&ag->ringing_work, K_NO_WAIT);
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_ALERTING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ACTIVE:
+		k_work_cancel_delayable(&ag->deferred_work);
+		break;
+	case BT_HFP_CALL_HOLD:
+		break;
+	default:
+		/* Invalid state */
+		break;
+	}
+
+	if (ag->state == BT_HFP_DISCONNECTING) {
+		int err = bt_rfcomm_dlc_disconnect(&ag->rfcomm_dlc);
+
+		if (err) {
+			LOG_ERR("Fail to disconnect DLC %p", &ag->rfcomm_dlc);
+		}
+	}
+}
+
+static void skip_space(struct net_buf *buf)
+{
+	while ((buf->len > 0) && (buf->data[0] == ' ')) {
+		(void)net_buf_pull(buf, 1);
+	}
+}
+
+static int get_number(struct net_buf *buf, uint32_t *number)
+{
+	int err = -EINVAL;
+
+	*number = 0;
+
+	skip_space(buf);
+	while ((buf->len > 0)) {
+		if ((buf->data[0] >= '0') && (buf->data[0] <= '9')) {
+			*number = *number * 10 + buf->data[0] - '0';
+			(void)net_buf_pull(buf, 1);
+			err = 0;
+		} else {
+			break;
+		}
+	}
+	skip_space(buf);
+
+	return err;
+}
+
+static int is_char(struct net_buf *buf, uint8_t c)
```

Comment:
It's quite confusing that the function has a boolean name, but is returning `int`. It should return `bool` (i.e. `true` instead of `0` and `false` instead of `-EINVAL`.

Reply:
Fixed.